ecllibrary/std/DataPatterns/Profile.ecl

```diff
@@ -514,6 +527,69 @@ EXPORT Profile(inFile,
         #UNIQUENAME(_MakeAttr);
         LOCAL %_MakeAttr%(STRING attr) := REGEXREPLACE('\\.', attr, '_');
 
+        // Determine if a UTF-8 string really contains UTF-8 characters
+        #UNIQUENAME(IsUTF8);
+        LOCAL BOOLEAN %IsUTF8%(DATA str) := EMBED(C++)
+            #option pure;
+
+            if (lenStr == 0)
+                return false;
+
+            const unsigned char*    bytes = reinterpret_cast<const unsigned char*>(str);
+            const unsigned char*    endPtr = bytes + lenStr;
+
+            while (bytes < endPtr)
+            {
+                if (bytes[0] == 0x09 || bytes[0] == 0x0A || bytes[0] == 0x0D || (0x20 <= bytes[0] && bytes[0] <= 0x7E))
+                {
+                    // ASCII; continue scan
+                    bytes += 1;
+                }
+                else if ((0xC2 <= bytes[0] && bytes[0] <= 0xDF) && (bytes+1 < endPtr) && (0x80 <= bytes[1] && bytes[1] <= 0xBF))
```

Comment:
If the parameter is a data you need to protect against accessing beyond the end of the buffer when you are looking at the last bytes of the string.

Reply:
That is the intended purpose of the newly-added (bytes+1 < endPtr) clause in this if().  With short-circuiting, that should prevent scanning beyond endPtr, no?

Reply:
Yes it does.  I should have paid closer attention to the change.